### PR TITLE
Use half traversal in FDBSCAN

### PR DESCRIPTION
**Pull request by aprokop**
Use half-traversal (#808) in the FDBSCAN algorithm.

### CODE REVIEW

**Comment by aprokop** [comment]
Two builds fails (`CUDA-11.0.3-Clang` and `SYCL`) because of
```
<snip>/src/ArborX_DBSCAN.hpp:72:12: error: no matching function for call to 'intersects'
    return intersects(ExperimentalHyperGeometry::Sphere<dim>{hyper_point, _r});
```
Compiling on Summit produces a more helpful message:
```
/ccs/home/aprokop/code/arborx/src/ArborX_DBSCAN.hpp(70): error: no instance of function template "ArborX::Details::intersects" matches the argument list
            argument types are: (ArborX::ExperimentalHyperGeometry::Sphere<3, float>)
```
So, for some reason, it is not trying `ArborX::intersects` in `ArborX_Predicates.hpp`, but the `ArborX::Details::intersects` from `ArborX_DetailsAlgorithms.hpp`.

What is particularly weird is that we have an almost identical code few lines lower, that does `attach(intersects(ExperimentalHyperGeometry::Sphere<dim>{hyper_point, w._r}));` and that one is just fine.

File: src/details/ArborX_DetailsFDBSCAN.hpp [hunk 1]
```diff
@@ -49,44 +49,48 @@ struct FDBSCANCallback
...
   CorePointsType _is_core_point;
 
-  template <typename Query>
-  KOKKOS_FUNCTION auto operator()(Query const &query, int j) const
+  KOKKOS_FUNCTION auto operator()(int i, int j) const
   {
-    int const i = ArborX::getData(query);
-
     bool const is_border_point = !_is_core_point(i);
+    bool const is_neighbor_core_point = _is_core_point(j);
```
Comment (dalg24) [inline src/details/ArborX_DetailsFDBSCAN.hpp:55]: I see it is old code but I still struggle with that name
Would `neighbor_is_core_point` be acceptable?

File: src/details/ArborX_DetailsFDBSCAN.hpp [hunk 2]
```diff
@@ -49,44 +49,48 @@ struct FDBSCANCallback
...
-    }
+      if (is_neighbor_core_point)
+      {
+        // For a border point that is connected to a core point, set its
+        // representative to that of the core point. If it is connected to
+        // multiple core points, it will be assigned to the cluster that the
+        // first found core point neighbor was in.
+        //
+        // NOTE: DO NOT USE merge(i, j) here. This may set this border point as
+        // a representative for the whole cluster potentially formbing a bridge
```
Comment (dalg24) [inline src/details/ArborX_DetailsFDBSCAN.hpp:66]:
```suggestion
        // a representative for the whole cluster potentially forming a bridge
```

File: src/details/ArborX_DetailsFDBSCAN.hpp [hunk 3]
```diff
@@ -49,44 +49,48 @@ struct FDBSCANCallback
...
-      _union_find.merge_into(j, i);
+      if (is_neighbor_core_point)
+      {
+        // For a core point that is connected to another core point, do the
+        // standard CCS algorithm
+        _union_find.merge(i, j);
+      }
+      else
+      {
+        // Merge the neighbor in (see comment above).
```
Comment (dalg24) [inline src/details/ArborX_DetailsFDBSCAN.hpp:85]: Be more specific.  See NOTE above?
Use a marker if necessary.

File: src/ArborX_DBSCAN.hpp [hunk 4]
```diff
@@ -55,6 +56,23 @@ struct PrimitivesWithRadius
...
+
+  template <typename Box>
+  KOKKOS_FUNCTION auto operator()(Box const &box) const
+  {
+    static_assert(GeometryTraits::is_box<Box>{});
+
+    constexpr int dim = GeometryTraits::dimension_v<Box>;
+    auto const &hyper_point =
+        reinterpret_cast<ExperimentalHyperGeometry::Point<dim> const &>(
+            box.minCorner());
```
Comment (dalg24) [inline src/ArborX_DBSCAN.hpp:71]: Why do you need the cast?

**Comment by aprokop** [comment]
## Crusher results

### HACC 497M (80M), eps = 0.014, minPts = 6
`./ArborX_Benchmark_DBSCAN.exe --binary --filename /gpfs/alpine/csc333/proj-shared/hacc_497M.arborx --max-num-points 80000000 --verbose --core-min-size 6 --eps 0.014 --impl fdbscan`

**master**
```
-- construction     :      0.108
-- query+cluster    :      2.812
---- neigh          :      0.204
---- query          :      2.592
-- postprocess      :      0.034
total time          :      2.965
```
**branch**
```
-- construction     :      0.108
-- query+cluster    :      1.572
---- neigh          :      0.203
---- query          :      1.353
-- postprocess      :      0.034
total time          :      1.725
```

### HACC 497M (80M), eps = 0.014, minPts = 2
`./ArborX_Benchmark_DBSCAN.exe --binary --filename /gpfs/alpine/csc333/proj-shared/hacc_497M.arborx --max-num-points 80000000 --verbose --core-min-size 2 --eps 0.014 --impl fdbscan`

**master**
```
-- construction     :      0.107
-- query+cluster    :      2.054
-- postprocess      :      0.037
total time          :      2.210
```
**branch**
```
-- construction     :      0.107
-- query+cluster    :      1.064
-- postprocess      :      0.038
total time          :      1.220
```

## Summit results

### HACC 497M (100M), eps = 0.014, minPts = 5
`./ArborX_Benchmark_DBSCAN.exe --binary --filename /gpfs/alpine/csc333/proj-shared/hacc_497M.arborx --max-num-points 100000000 --verbose --core-min-size 5 --eps 0.014 --impl fdbscan`

**master**
```
-- construction     :      0.274
-- query+cluster    :      3.485
---- neigh          :      0.196
---- query          :      3.239
-- postprocess      :      0.065
total time          :      3.835
```
**branch**
```
-- construction     :      0.274
-- query+cluster    :      2.304
---- neigh          :      0.196
---- query          :      2.058
-- postprocess      :      0.065
total time          :      2.653
```

### HACC 37M, eps = 0.042, minPts = 2
`./ArborX_Benchmark_DBSCAN.exe --binary --filename /gpfs/alpine/csc333/proj-shared/hacc_37M.arborx --verbose --core-min-size 2 --eps 0.042 --impl fdbscan
`
**master**
```
-- construction     :      0.103
-- query+cluster    :      0.288
-- postprocess      :      0.018
total time          :      0.412
```
**branch**
```
-- construction     :      0.103
-- query+cluster    :      0.167
-- postprocess      :      0.018
total time          :      0.292

**Review by dalg24 — APPROVED** [review]
This is excellent.  Please make sure you update the description of the PR before you merge.
(There is no description at the moment)

**Pull request merged**